File: emulinker/src/main/java/org/emulinker/kaillera/controller/v086/protocol/ClientACK.java
package org.emulinker.kaillera.controller.v086.protocol;

import java.nio.ByteBuffer;
import org.emulinker.kaillera.controller.messaging.*;
import org.emulinker.util.*;

public class ClientACK extends ACK {
  public static final byte ID = 0x06;
  public static final String DESC = "Client to Server ACK";

  public ClientACK(int messageNumber) throws MessageFormatException {
    super(messageNumber, 0, 1, 2, 3);
  }

  @Override
  public byte getID() {
    return ID;
  }

  @Override
  public String getDescription() {
    return DESC;
  }

  public static ClientACK parse(int messageNumber, ByteBuffer buffer)
      throws ParseException, MessageFormatException {
    if (buffer.remaining() < 17) throw new ParseException("Failed byte count validation!");

    byte b = buffer.get();

    if (b != 0x00) {
      throw new MessageFormatException(
          "Invalid " + DESC + " format: byte 0 = " + EmuUtil.byteToHex(b));
    }

    // long val1 = UnsignedUtil.getUnsignedInt(buffer);
    // long val2 = UnsignedUtil.getUnsignedInt(buffer);
    // long val3 = UnsignedUtil.getUnsignedInt(buffer);
    // long val4 = UnsignedUtil.getUnsignedInt(buffer);

    // if (val1 != 0 || val2 != 1 || val3 != 2 || val4 != 3)
    // throw new MessageFormatException("Invalid " + DESC + " format: bytes do not match acceptable
    // format!");

    return new ClientACK(messageNumber);
  }
}

File: emulinker/src/main/java/org/emulinker/kaillera/controller/v086/protocol/ClientACK.kt
package org.emulinker.kaillera.controller.v086.protocol

import java.nio.ByteBuffer
import kotlin.Throws
import org.emulinker.kaillera.controller.messaging.MessageFormatException
import org.emulinker.kaillera.controller.messaging.ParseException
import org.emulinker.util.EmuUtil

data class ClientACK
    @Throws(MessageFormatException::class)
    constructor(override val messageNumber: Int) : ACK() {

  override val val1: Long = 0
  override val val2: Long = 1
  override val val3: Long = 2
  override val val4: Long = 3

  override val messageId = ID

  init {
    validateMessageNumber(messageNumber)
  }

  companion object {
    const val ID: Byte = 0x06

    @Throws(ParseException::class, MessageFormatException::class)
    fun parse(messageNumber: Int, buffer: ByteBuffer): ClientACK {
      if (buffer.remaining() < 17) throw ParseException("Failed byte count validation!")
      val b = buffer.get()
      if (b.toInt() != 0x00) {
        throw MessageFormatException(
            "Invalid " + "Client to Server ACK" + " format: byte 0 = " + EmuUtil.byteToHex(b))
      }

      // long val1 = UnsignedUtil.getUnsignedInt(buffer);
      // long val2 = UnsignedUtil.getUnsignedInt(buffer);
      // long val3 = UnsignedUtil.getUnsignedInt(buffer);
      // long val4 = UnsignedUtil.getUnsignedInt(buffer);

      // if (val1 != 0 || val2 != 1 || val3 != 2 || val4 != 3)
      // throw new MessageFormatException("Invalid " + DESC + " format: bytes do not match
      // acceptable
      // format!");
      return ClientACK(messageNumber)
    }
  }
}
